src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -344,6 +361,43 @@ public void onFailure(Throwable e) {
         }
     }
 
+    protected boolean shardsAllocatedOnLocalNode(ClusterState state, IndexMetaData indexMetaData) {
+        boolean shardsAllocatedOnThisNode = false;
+        IndexRoutingTable indexRoutingTable = state.getRoutingTable().index(indexMetaData.index());
+        if (indexRoutingTable == null) {
+            // nothing allocated ?
+            return false;
+        }
+        // iterate over shards and see if one is on our node
+        for (IntObjectCursor it : indexRoutingTable.shards()) {
+            IndexShardRoutingTable shardRoutingTable = (IndexShardRoutingTable) it.value;
+            for (ShardRouting shardRouting : shardRoutingTable.shards()) {
+                if (shardRouting.currentNodeId() != null && shardRouting.currentNodeId().equals(state.nodes().localNode().getId())) {
+                    shardsAllocatedOnThisNode = true;
+                }
+            }
+        }
+        return shardsAllocatedOnThisNode;
+    }
+
+    protected boolean isDataOnlyNode(ClusterState state) {
+        return ((state.nodes().localNode().masterNode() == false) && (state.nodes().localNode().dataNode() == true));
+    }
+
+    private void removeIndexState(IndexMetaData indexMetaData) {
```

Comment:
this method must go away. We do index cleaning in IndicesService once indices are deleted. If it's buggy please open a different issue